packages/flutter/lib/src/widgets/platform_menu_bar.dart

```diff
@@ -0,0 +1,886 @@
+/// serialization of a [DefaultPlatformMenuDelegateSerializable].
+typedef DefaultPlatformMenuDelegateSerializableIdGenerator = int Function(MenuItem item);
+
+/// A mixin for doing serialization of [MenuItem] objects.
+///
+/// This is used by the [DefaultPlatformMenuDelegate] to serialize menu
+/// hierarchies for sending to the platform for rendering.
+mixin DefaultPlatformMenuDelegateSerializable {
+  /// Converts the representation of this item into a map suitable for sending
+  /// over the default "flutter/menu" channel used by
+  /// [DefaultPlatformMenuDelegate].
+  ///
+  /// The `delegate` is a [DefaultPlatformMenuDelegate] that is requesting the
+  /// serialization. The `index` is the position of this menu item in the list
+  /// of children of the [PlatformMenu] it belongs to, and `count` is the number
+  /// of children in the [PlatformMenu] it belongs to.
+  Iterable<Map<String, Object?>> toChannelRepresentation(
+    DefaultPlatformMenuDelegate delegate, {
+    required int index,
+    required int count,
+    required DefaultPlatformMenuDelegateSerializableIdGenerator getId,
+  });
+}
+
+/// The platform menu delegate that handles the built-in macOS platform menu
+/// generation using the 'flutter/menu' channel.
+///
+/// An instance of this class is set on [WidgetsBinding.platformMenuDelegate] by
+/// default when the [WidgetsBinding] is initialized.
+///
+/// See also:
+///
+///  * [PlatformMenuBar], the widget that adds a platform menu bar to an
+///    application.
+///  * [PlatformMenu], the class that describes a menu item with children
+///    that appear in a cascading menu.
+///  * [PlatformMenuItem], the class that describes the leaves of a menu
+///    hierarchy.
+class DefaultPlatformMenuDelegate extends PlatformMenuDelegate {
+  /// Creates a const [DefaultPlatformMenuDelegate].
+  ///
+  /// The optional [channel] argument defines the channel used to communicate
+  /// with the platform. It defaults to [SystemChannels.menu] if not supplied.
+  DefaultPlatformMenuDelegate({MethodChannel? channel})
+      : channel = channel ?? SystemChannels.menu,
+        _idMap = <int, MenuItem>{} {
+    this.channel.setMethodCallHandler(_methodCallHandler);
+  }
+
+  // Map of distributed IDs to menu items.
+  final Map<int, MenuItem> _idMap;
+  // An ever increasing value used to dole out IDs.
+  int _serial = 0;
+  // The context used to "lock" this delegate to a specific instance of
+  // PlatformMenuBar to make sure there is only one.
+  BuildContext? _lockedContext;
+
+  @override
+  void clearMenus() => setMenus(<MenuItem>[]);
+
+  @override
+  void setMenus(List<MenuItem> topLevelMenus) {
+    _idMap.clear();
+    final List<Map<String, Object?>> representation = <Map<String, Object?>>[];
+    if (topLevelMenus.isNotEmpty) {
+      int index = 0;
+      for (final MenuItem childItem in topLevelMenus) {
+        if (childItem is! DefaultPlatformMenuDelegateSerializable) {
```

Comment:
I still find this runtime type checking odd. It's not a pattern we generally use in the framework. Could the `PlatformMenuDelegate` just take `SerializableMenuItem` that just have serialization as part of their interface? Is there a use case where you want a custom `PlatformMenuDelegate` that doesn't need to serialize the provided menu?